gps/ublox.go

```diff
@@ -1,53 +1,227 @@
+			return err
+		}
+		if err = SendCommand(d, d.buffer[:]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GNSSDisableCmd is a UBX-CFG-GNSS command to disable all GNSS but GPS
+// Needed for MAX8's, not needed for MAX7
+var GNSSDisableCmd = CfgGnss{
+	MsgVer:      0x00,
+	NumTrkChHw:  0x20, // 32 channels
+	NumTrkChUse: 0x20,
+	ConfigBlocks: []*CfgGnssConfigBlocksType{
+		{GnssId: 0, ResTrkCh: 8, MaxTrkCh: 16, Flags: CfgGnssEnable | 0x010000}, // GPS enabled
+		{GnssId: 1, ResTrkCh: 1, MaxTrkCh: 3, Flags: 0x010000},                  // SBAS disabled
+		{GnssId: 3, ResTrkCh: 8, MaxTrkCh: 16, Flags: 0x010000},                 // BeiDou disabled
+		{GnssId: 5, ResTrkCh: 0, MaxTrkCh: 3, Flags: 0x010000},                  // QZSS disabled
+		{GnssId: 6, ResTrkCh: 8, MaxTrkCh: 14, Flags: 0x010000},                 // GLONASS disabled
+	},
+}
+
+// SetGNSSDisable sends UBX-CFG-GNSS command to disable all GNSS but GPS
+func SetGNSSDisable(d Device) (err error) {
+	if _, err = GNSSDisableCmd.Write(d.buffer[:]); err != nil {
+		return err
+	}
+	return SendCommand(d, d.buffer[:])
+}
+
+// SendCommand sends a UBX command and waits for ACK/NAK response
+func SendCommand(d Device, command []byte) error {
+	// Calculate and append checksum
+	checksummed := appendChecksum(command)
+	d.WriteBytes(checksummed)
+
 	start := time.Now()
-	for time.Now().Sub(start) < 1000 {
-		if d.readNextByte() == '\n' {
-			if d.readNextByte() == 0xB5 {
-				d.readNextByte()
-				if d.readNextByte() == 0x05 {
-					if d.readNextByte() == 0x01 {
-						return
-					}
-				}
-			}
-		}
-	}
-	return errors.New("no ACK to GPS command")
+	for time.Since(start) < time.Second {
+		// Look for UBX sync sequence
+		if d.readNextByte() != ubxSyncChar1 {
+			continue
+		}
+		if d.readNextByte() != ubxSyncChar2 {
+			continue
+		}
+
+		// Read message class and ID
+		msgClass := d.readNextByte()
+		msgID := d.readNextByte()
+
+		// Check if it's an ACK class message
+		if msgClass != ubxClassACK {
+			continue
+		}
+
+		// Read length (2 bytes, little-endian) - ACK is always 2 bytes payload
+		lenLo := d.readNextByte()
+		lenHi := d.readNextByte()
+		length := uint16(lenLo) | uint16(lenHi)<<8
+
+		if length != 2 {
+			continue
+		}
+
+		// Read ACK payload: class and ID of acknowledged message
+		ackClass := d.readNextByte()
+		ackID := d.readNextByte()
+
+		// Verify ACK is for our command (command[2] = class, command[3] = ID)
+		if ackClass != command[2] || ackID != command[3] {
+			continue
+		}
+
+		if msgID == ubxACK_ACK {
+			return nil
+		}
+		if msgID == ubxACK_NAK {
+			return errGPSCommandRejected
+		}
+	}
+
+	return errNoACKToGPSCommand
+}
+
+// appendChecksum calculates UBX checksum and appends it to the message
+func appendChecksum(msg []byte) []byte {
+	var ckA, ckB byte
+	// Checksum covers class, ID, length, and payload (skip sync chars)
+	for i := 2; i < len(msg); i++ {
+		ckA += msg[i]
+		ckB += ckA
+	}
+	return append(msg, ckA, ckB)
```

Comment:
This gonna heap allocate like crazy, gonna not? Probably fine in the supposed application?

Reply:
@ysoldak see additional commit that removed the heap allocation on error return.

```
$ tinygo build -target pico -size short -print-allocs=gps ./examples/gps/uart/
/home/ron/Development/tinygo/drivers/gps/gps.go:46:17: object allocated on the heap: escapes at line 46
   code    data     bss |   flash     ram
  68372    1640    5416 |   70012    7056
```

Now only the allocation for the buffer remains.